File: java/linkedlist/Queue.java
// Definition of singly-linked list:

// public class ListNode {
//     int val;
//     ListNode next;

//     public ListNode(int val) {
//         this.val = val;
//         this.next = null;
//     }
// }

public class Queue {
    ListNode left;  // front of Queue   front -> [1,2,3]
    ListNode right; // back of Queue   [1,2,3] <- back
    
    public Queue() {
        this.left  = null;
        this.right = null;
    }

    public void enqueue(int val) {
        ListNode newNode = new ListNode(val);
        if (this.right != null) {  
        // Queue is not empty 
            this.right.next = newNode;
            this.right = this.right.next;
        } else {       
        // Queue is empty             
            this.left = newNode;
            this.right = newNode;
        }
    }

    public int dequeue() {
        if (this.left == null) {
        // Queue is empty 
            System.exit(0);
        }
        int val = this.left.val;
        this.left = this.left.next;
        return val;
        
    }

    public void print() {
        ListNode cur = this.left;
        while(cur != null) {
            System.out.print(cur.val + " -> ");
            cur = cur.next;
        }
        System.out.println();
    }

}

// public class Main {
//     public static void main(String[] args) {
//         Queue q = new Queue();

//         q.enqueue(1);  // 1  ->
//         q.enqueue(2);  // 1  -> 2  ->
       
//         q.dequeue();   // 2 ->
    
//         q.enqueue(3);  // 2  -> 3 ->
//         q.enqueue(4);  // 2  -> 3 -> 4 ->
       
//         q.dequeue();   // 3  -> 4 ->
//         q.dequeue();   // 4 ->
       
//     }
// }

File: csharp/linkedlist/Queue.cs
// public class ListNode {
//     int Value;
//     ListNode Next;
//
//     public ListNode(int val) {
//         this.Value = val;
//         this.Next = null;
//     }
// }

public class Queue
{
    ListNode Left;  // front of Queue   front -> [1,2,3]
    ListNode Right; // back of Queue   [1,2,3] <- back

    public Queue()
    {
        this.Left = null;
        this.Right = null;
    }

    public void EnQueue(int val)
    {
        ListNode newNode = new ListNode(val);
        if (this.Right != null)
        {
            // Queue is not empty 
            this.Right.Next = newNode;
            this.Right = this.Right.Next;
        }
        else
        {
            // Queue is empty             
            this.Left = newNode;
            this.Right = newNode;
        }
    }

    public int DeQueue()
    {
        if (this.Left == null)
        {
            // Queue is empty 
            return -1;
        }
        int val = this.Left.Value;
        this.Left = this.Left.Next;
        if (this.Left == null)
        {
            this.Right = null;
        }
        return val;
    }

    public void Print()
    {
        ListNode cur = this.Left;
        while (cur != null)
        {
            Console.WriteLine(cur.Value + " -> ");
            cur = cur.Next;
        }
        Console.WriteLine("");
    }

}
